cilium-cli/connectivity/tests/encryption_v2.go

```diff
@@ -0,0 +1,592 @@
+// to pod traffic for the client and server pods under test.
+//
+// For tunnel mode the TCPDUMP filter looks for UDP packets which match:
+// 1. UDP protocol
+// 2. UDP dst port matching the configured tunnel mode's well known port (e.g. 8472 for VXLAN)
+// 3. Inner IP headers of pod-to-pod traffic.
+// In other words the TCPDUMP filter seeks into the tunnel packet and looks for
+// plain-text pod-to-pod traffic for the client and server pods under test.
+//
+// Leak detection is performed both on the client side and server side, ensuring
+// client->server traffic is encrypteda and the return traffic, server->client
+// is encrypted as well.
+//
+// This test should be ran both when encryption is enabled AND disabled.
+//
+// When encryption is enabled the test will check that no packets match the
+// TCPDUMP filter, i.e. there are no plain text leaks
+//
+// When encryption is disabled the test will check that packets DO match the
+// TCPDUMP filter, this is a sanity check to ensure we match plain-text packets
+// appropriately and have confidence that aforementioned leak detection works.
+func PodToPodEncryptionV2() check.Scenario {
+	return &podToPodEncryptionV2{}
+}
+
+type podToPodEncryptionV2 struct {
+	ct *check.ConnectivityTest
+	// client pod used to generate traffic
+	client *check.Pod
+	// server pod which receives and responds to client traffic
+	server *check.Pod
+	// pod on client's node providing access to host network namespace
+	clientHostNS *check.Pod
+	// pod on server's node providing access to host network namespace
+	serverHostNS *check.Pod
+
+	tunnelMode  features.Status
+	encryptMode features.Status
+	ipv4Enabled features.Status
+	ipv6Enabled features.Status
+
+	// the egress device, on the client node, that client->server traffic will
+	// leave on.
+	//
+	// NOTE: this test assumes that if IPv6 is enabled the same egress device
+	// is used to push client traffic toward the server.
+	// this will almost always be the case.
+	clientEgressDev string
+	// the egress device, on the server node, that server->client (return traffic)
+	// will leave on.
+	//
+	// see clientEgressDev NOTE:
+	serverEgressDev string
+
+	// pcap filter used to detect leaks on the client side
+	clientFilter4 string
+	// pcap filter used to detect leaks on the server side
+	serverFilter4 string
+	// tcpdump running on the client
+	clientSniffer4 *sniff.Sniffer
+	// tcpdump running on the server
+	serverSniffer4 *sniff.Sniffer
+
+	// IPv6 variants of the above
+	clientFilter6  string
+	serverFilter6  string
+	clientSniffer6 *sniff.Sniffer
+	serverSniffer6 *sniff.Sniffer
+}
+
+func (s *podToPodEncryptionV2) Name() string {
+	return "pod-to-pod-encryption-v2"
+}
+
+// resolveEgressDevice resolves the egress device used in the provided host
+// network namespace used to send traffic to dst.
+func (s *podToPodEncryptionV2) resolveEgressDevice(ctx context.Context, srcHostNS *check.Pod, dst *check.Pod) (string, error) {
```

Comment:
How about re-using the `github.com/vishvananda/netlink` package? IIUC, rather than manually parsing the output, we could just do something like:

```go
func getRoutingDevice(ip string) (string, error) {
	// Parse the IP address
	route, err := netlink.RouteGet(net.ParseIP(ip))
	if err != nil || len(route) == 0 {
		return "", fmt.Errorf("no route found for IP %s", ip)
	}

	// Get the link (device) associated with the route
	link, err := netlink.LinkByIndex(route[0].LinkIndex)
	if err != nil {
		return "", err
	}

	return link.Attrs().Name, nil
}
```

Reply:
This would be nice, but the cilium-cli most likely doesn't run anywhere near the network namespace we are targeting. It may but ever be on the same machine.

Reply:
Uooops you're right, thanks for reminding!